### PR TITLE
feat: Contract subscriptions filters

### PR DESCRIPTION
**Pull request by arcoraven**
- Allow CRUD for contract subscription filters
- What are filters? The dev can select whether or not to enable logs or receipts.
  - For logs, they can listen to all events for specific events.
  - For receipts, they can listen to all functions or specific functions.

---

## PR-Codex overview
The focus of this PR is to improve consistency and readability in component names, variable casing, and feature enhancements.

### Detailed summary
- Updated casing for server and database names
- Improved filtering options for contract subscriptions
- Added a new hook for fetching contract ABI items
- Enhanced UI for adding contract subscriptions in the modal

> The following files were skipped due to too many changes: `src/components/engine/contract-subscription/add-contract-subscription-button.tsx`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: src/hooks/useContractAbiItems.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+/**
+ * Returns read/write function and event names from a contract.
+ *
+ * @param chainId number
+ * @param address Address
+ * @returns readFunctions - List of read function names
+ * @returns writeFunctinos - List of write function names
+ * @returns events - List of event names
+ */
+export const useContractAbiItems = (
```
Comment (arcoraven) [inline src/hooks/useContractAbiItems.ts:16]: I'm not familiar with writing hooks. Please look at this.

**Comment by arcoraven** [comment]
Some screenshots

image
image
image
image
image
image
image
image

File: src/hooks/useContractAbiItems.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        console.error(
+          `Failed to load ABI for contract ${address} on chain ${chainId}.`,
+        );
+        setWriteFunctions([]);
+        setEvents([]);
+      }
+    })();
+  }, [chainId, address]);
+
+  return { readFunctions, writeFunctions, events };
```
Comment (jnsdls) [inline src/hooks/useContractAbiItems.ts:74]: this entire thing should be a `useQuery` instead: `chainId` and `contractAddress` should be part of the query key

something along these lines (rough):
```
const {data, isLoading} = useQuery({
  queryKey: ['contract-abi-items', chainId, address],
  queryFn: async () => {
// ... your async logic here, return the readFunctions, writeFunctions, events from it at the end
})

return data;
```

File: src/components/engine/contract-subscription/add-contract-subscription-button.tsx [hunk 3]
```diff
@@ -72,21 +89,34 @@ const AddModal = ({
...
+      processEventLogs: true,
+      filterEvents: [],
+      processTransactionReceipts: false,
+      filterFunctions: [],
     },
   });
 
-  const onSubmit = (data: AddModalInput) => {
+  const onSubmit = () => {
+    const formData = form.getValues();
```
[the remark this answers is not in the document]
Comment (arcoraven) [reply src/components/engine/contract-subscription/add-contract-subscription-button.tsx:110]: This modal has two pages/steps, so I wasn't sure if wrapping the modal in a form made sense. I guess it does.

**Comment by jnsdls** [comment]
* **#2653** <a href="https://app.graphite.dev/github/pr/thirdweb-dev/dashboard/2653?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @arcoraven and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/components/engine/contract-subscription/add-contract-subscription-button.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (jnsdls) [reply src/components/engine/contract-subscription/add-contract-subscription-button.tsx:110]: yeah it does

**Review by jnsdls — CHANGES_REQUESTED** [review]
fixed the useQuery - just the form left

**Review by jnsdls — APPROVED** [review]
LGTM

**Pull request merged**